jib-core/src/main/java/com/google/cloud/tools/jib/tar/TarStreamBuilder.java

```diff
@@ -56,45 +48,35 @@ private void writeEntriesAsTarArchive(OutputStream tarByteStream) throws IOExcep
         new TarArchiveOutputStream(tarByteStream, StandardCharsets.UTF_8.name())) {
       // Enables PAX extended headers to support long file names.
       tarArchiveOutputStream.setLongFileMode(TarArchiveOutputStream.LONGFILE_POSIX);
-      for (Map.Entry<TarArchiveEntry, TarArchiveOutputStreamConsumer> entry :
-          archiveMap.entrySet()) {
+      for (Map.Entry<TarArchiveEntry, ByteSource> entry : archiveMap.entrySet()) {
         tarArchiveOutputStream.putArchiveEntry(entry.getKey());
-        entry.getValue().accept(tarArchiveOutputStream);
+        tarArchiveOutputStream.write(entry.getValue().read());
         tarArchiveOutputStream.closeArchiveEntry();
       }
     }
   }
 
   /**
-   * Adds an entry to the archive.
+   * Adds a {@link TarArchiveEntry} to the archive.
    *
-   * @param entry the entry to add.
+   * @param entry the {@link TarArchiveEntry}
    */
   public void addEntry(TarArchiveEntry entry) {
     archiveMap.put(
         entry,
-        tarArchiveOutputStream -> {
-          // Note that this will skip files that don't exist.
-          if (entry.isFile()) {
-            try (InputStream contentStream =
-                new BufferedInputStream(Files.newInputStream(entry.getFile().toPath()))) {
-              ByteStreams.copy(contentStream, tarArchiveOutputStream);
-            }
-          }
-        });
+        entry.isFile() ? MoreFiles.asByteSource(entry.getFile().toPath()) : ByteSource.empty());
   }
 
   /**
    * Adds a blob to the archive.
    *
-   * @param contents the blob contents to add to the tarball.
-   * @param name the name of the entry (i.e. filename).
+   * @param contents the blob contents to add to the tarball
+   * @param name the name of the entry (i.e. filename)
    */
-  public void addEntry(String contents, String name) {
+  public void addEntry(byte[] contents, String name) {
```

Comment:
I think the point of using `String` as a parameter type is to prevent callers from accidentally using this signature for files, where they read the file into a `byte[]` and then call this method.

Reply:
How does using `String` instead of `byte[]` prevent that?

Reply:
There's a `Files.readAllBytes` method that is used to read files to a `byte[]` and not a method for reading to string - and reading files to a string in general is not a good idea unless that file is guaranteed to be a text file.

Reply:
I'd argue that making it a `String` doesn't help much with that, considering you can just wrap the `Files.readAllBytes` in a `new String()` (which we actually used to do in the [test](https://github.com/GoogleContainerTools/jib/pull/689/files/52c6e480812342d79bc4e29bd8c6436c3dc262c5#diff-843239071ea940328c65b2c1f0d2ba94L58)). Might just be more useful to update the javadoc to warn against using this signature with files.

Reply:
That sounds fine (updating the javadoc), although I'd argue that generally, we should keep APIs as type-restricted as possible (ie. favor `SpecificType` over `Object` and `DataType` over `byte[]`).

Reply:
Perhaps `JsonTemplate` might be the more applicable type here instead of `String` - although that introduces a dependency from the `tar` package to the `json` package.

Reply:
> `String` is encoded by `UTF-8`

I think I wasn't really correct with this statement. https://docs.oracle.com/javase/6/docs/api/java/lang/Character.html#unicode says

> The Java 2 platform uses the UTF-16 representation in `char` arrays and in the `String` and `StringBuffer` classes.

Anyways, it doesn't matter regarding converting a `String` to bytes with any encoding; `getBytes(Charset)` will do the righr job. So, I guess the question here is, does the TAR stream always expect that text strings should be written in the UTF-8 encoding? That is, if someone reads the written bytes later, will it always decode the bytes with UTF-8? If the answer is yes, I *think* it is not so wrong to accept `String` (which should be plain texts, by the way).

Reply:
Perhaps the right solution here is to change the signature to `public void addEntry(String jsonString, String name)`, since JSON is defined to be encoded in UTF-8 [[RFC8259](https://tools.ietf.org/html/rfc8259#section-8.1)],

Reply:
@elharo Would be interested in hearing what you think.

Reply:
@briandealwis as well

Reply:
I don't know if the TAR string entry specification mandates writing strings in UTF-8, but if it is in fact true, I'd be OK with something like `addTextEntryInUtf8(String text, String name)`.

Reply:
As tar files are just bags of bytes, I'd be happy with using `Blob`.  It has the advantage of minimizing Guava types in our method signatures.  Let the caller determine how to encode whatever they're passing in.

Reply:
I agree - `Blob` would be good too as we have a purpose-built API for our use cases for that.

Reply:
I guess the direction we're headed with this class is less a general tarball builder and more something to fit our specific purposes, and the only things we're ever adding are files and raw strings. I'll stick with the string parameter and just rename the method as @chanseokoh suggested.